### PR TITLE
implement grabElementBoundingRect for wdio and puppeteer helpers

### PR DESCRIPTION
**Pull request by PeterNgTr**
https://github.com/Codeception/CodeceptJS/issues/1867
this would add `grabElementBoundingRect`method for WDIO and Puppeteer.

### CODE REVIEW

File: lib/helper/WebDriver.js [hunk 1]
```diff
@@ -2055,6 +2055,17 @@ class WebDriver extends Helper {
...
 
+  /**
+   * {{> grabElementSize }}
+   *
+   */
+  async grabElementSize(locator, prop) {
+    const res = await this._locate(locator, true);
+    assertElementExists(res, locator);
+
+    return forEachAsync(res, async el => el.getSize(prop));
```
Comment (Vorobeyko) [inline lib/helper/WebDriver.js]: Webdriver return only width and height, but Puppeteer method return x, y, width and height.
Can you make the same return values?

Comment (PeterNgTr) [reply lib/helper/WebDriver.js]: Sure, not a problem :)

Comment (PeterNgTr) [reply lib/helper/WebDriver.js]: @Vorobeyko thanks bro for your review! I pushed the changes!

File: lib/helper/Puppeteer.js [hunk 2]
```diff
@@ -1985,6 +1985,31 @@ class Puppeteer extends Helper {
...
+    const result = [];
+    els = els.filter(v => v);
+
+    if (prop) {
+      els.forEach((element) => {
+        result.push(element[prop]);
+      });
+    } else {
+      els.forEach((element) => {
+        delete element.x;
```
Comment (Vorobeyko) [inline lib/helper/Puppeteer.js]: Oh no) 
x and y can be use! My command used this property)) 
I thought you make x and y in webdriverio)))

Comment (Vorobeyko) [reply lib/helper/Puppeteer.js]: X and y coordinates are important))

Comment (PeterNgTr) [reply lib/helper/Puppeteer.js]: haha. I'm thinking of renaming this method as `grabElementInfo` to convey better meaning. Cause we want to return the element size and location(x,y). Do you have any suggestions?

Comment (Vorobeyko) [reply lib/helper/Puppeteer.js]: Maybe grabElementBoundingRect ?

Comment (PeterNgTr) [reply lib/helper/Puppeteer.js]: sounds nice. I give a go for it.

Comment (PeterNgTr) [reply lib/helper/Puppeteer.js]: @Vorobeyko done. not both wdio and puppeteer return location, weight and height. :)

File: lib/helper/Puppeteer.js [hunk 3]
```diff
@@ -1985,6 +1985,29 @@ class Puppeteer extends Helper {
   async grabDataFromPerformanceTiming() {
     return perfTiming;
   }
+
+  /**
+   * {{> grabElementSize }}
```
Comment (Vorobeyko) [inline lib/helper/Puppeteer.js]: grabElementBoundingRect

File: lib/helper/Puppeteer.js [hunk 4]
```diff
@@ -1985,6 +1985,29 @@ class Puppeteer extends Helper {
...
+   *
+   */
+  async grabElementBoundingRect(locator, prop) {
+    let els = await this._locate(locator);
+    els = await Promise.all(els.map(el => el.boundingBox()));
+    const result = [];
+    els = els.filter(v => v);
```
Comment (Vorobeyko) [inline lib/helper/Puppeteer.js]: what this line do?

File: lib/helper/WebDriver.js [hunk 5]
```diff
@@ -2055,6 +2055,34 @@ class WebDriver extends Helper {
...
+      arr.forEach((element) => {
+        result.push(element[prop]);
+      });
+    } else {
+      arr.forEach((element) => {
+        result.push(element);
+      });
+    }
+
+    return result;
```
Comment (Vorobeyko) [inline lib/helper/WebDriver.js]: You can use `find` for find element with prop and immediately return the result, like
```
if (prop) {
      return [arr.find((element) => {...});]
    }
```

If prop is null - return els.

File: lib/helper/Puppeteer.js (hunk 4 above)
Comment (PeterNgTr) [reply lib/helper/Puppeteer.js]: this line to remove empty items, null, etc.

File: docs/webapi/grabElementBoundingRect.mustache [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+```js
+const value = await I.grabElementBoundingRect('h3');
+const value = await I.grabElementBoundingRect('h3', 'width');
+```
+
+If multiple elements found returns an array of texts.
+
+@param {string|object} locator element located by CSS|XPath|strict locator.
+@param {string} elementSize the width or height of the given element.
+@returns {Promise<array>} Element size
```
Comment (DavertMik) [inline docs/webapi/grabElementBoundingRect.mustache]: The structure is not defined here. I would like to see which object is returned

File: docs/webapi/grabElementBoundingRect.mustache [hunk 7]
```diff
@@ -0,0 +1,14 @@
+Grab the width, height, location of given locator.
+Provide `width` or `height`as second param to get your desired prop.
+Resumes test execution, so **should be used inside an async function with `await`** operator.
+
+```js
+const value = await I.grabElementBoundingRect('h3');
```
Comment (DavertMik) [inline docs/webapi/grabElementBoundingRect.mustache]: please add example on using `value`, show which properties it includes.

File: docs/webapi/grabElementBoundingRect.mustache [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+const value = await I.grabElementBoundingRect('h3', 'width');
+```
+
+If multiple elements found returns an array of texts.
```
Comment (DavertMik) [inline docs/webapi/grabElementBoundingRect.mustache]: I know that we have this feature for `grabTextFrom` but I think it was a design mistake. 
Please make this return an object, not an array, to make the behavior more predictable.

**Comment by DavertMik** [comment]
Sorry, @PeterNgTr I was unclear in the last statement:

> I know that we have this feature for grabTextFrom but I think it was a design mistake.
Please make this return an object, not an array, to make the behavior more predictable.

What I actually meant, this method should always return an object for a ONE element. Not an array.
So this method should take the first matched element and returns an object. If you need a specific object in a list - use `locate().at()`. 

If you don't mind I will merge this PR but change this method to return rect only for the first element

**Comment by PeterNgTr** [comment]
@davertk please go with your suggestion.

**Comment by DavertMik** [comment]
Great! Doing so. If you may need multipleRect this should be additional method... However, I hope you won't need it :smile:

**Pull request merged**